torch/csrc/distributed/rpc/process_group_agent.cpp

```
@@ -319,46 +330,65 @@ void ProcessGroupAgent::enqueueSend(SendWork work) {
+            pendingSends.reserve(1);
+            {
+              std::lock_guard<std::mutex> guard(sendMutexes_[dst]);
+              pendingSends.emplace_back(
+                  pg_->send(preamble, dst, dst /* channelTag */));
+            }
+          } else {
+            std::vector<torch::Tensor> payload = {torch::from_blob(
+                (void*)serializedPayload.c_str(),
+                serializedPayload.length(),
+                {torch::kChar})};
+            pendingSends.reserve(2);
+
+            sendCounts_.increment(dst);
+
+            {
+              std::lock_guard<std::mutex> guard(sendMutexes_[dst]);
+              pendingSends.emplace_back(
+                  pg_->send(preamble, dst, dst /* channelTag */));
+              pendingSends.emplace_back(
+                  pg_->send(payload, dst, dst /* channelTag */));
+            }
           }
-        } else {
-          std::vector<torch::Tensor> payload = {torch::from_blob(
-              (void*)serializedPayload.c_str(),
-              serializedPayload.length(),
-              {torch::kChar})};
-          pendingSends.reserve(2);
-
-          sendCounts_.increment(dst);
-
-          {
-            std::lock_guard<std::mutex> guard(sendMutexes_[dst]);
-            pendingSends.emplace_back(
-                pg_->send(preamble, dst, dst /* channelTag */));
-            pendingSends.emplace_back(
-                pg_->send(payload, dst, dst /* channelTag */));
+          for (auto& pendingSend : pendingSends) {
+            pendingSend->wait();
+          }
+        } catch (std::exception& e) {
+          if (work.message_.isRequest()) {
+            std::ostringstream ss;
+            ss << "Encountered exception in ProcessGroupAgent::enqueueSend: "
+               << e.what();
+            auto exceptionMsg =
+                rpc::createExceptionResponse(work.message_, ss.str());
+            markFutureWithMessage(exceptionMsg);
+          }
+        } catch (...) {
```

wondering: do we ever expect to see something that's not an std::exception?
if not, let's just drop this code - let it crash, and we can keep the code simpler and fix up as needed.

```
If we do, maybe we unify the codepaths as the follows:
   try {
      // send code
    } catch (std::exception& e) {
        handleSendFailure("Encountered sending error " + e.what(), work.message_);
    } catch (...) {
        handleSendFailure("Encountering unknown send error", work.messaage_);
    }
```

---

I don't see this currently and can't think of a case where we'd throw something that's not an std::exception, as that'd be very surprising, so it might be better to crash here per your suggestion.